chpl/chpl-service/src/main/java/gov/healthit/chpl/dao/statistics/ChartDataDAOImpl.java

```diff
@@ -0,0 +1,183 @@
+			} else {
+				entity.setLastModifiedDate(new Date());
+			}
+			
+			create(entity);
+			return new ChartDataDTO(entity);
+		}
+	}
+	
+	@Transactional
+	public ChartDataDTO update(ChartDataDTO dto) throws EntityRetrievalException{
+		
+		ChartDataEntity entity = getEntityById(dto.getId());	
+		if(entity == null) {
+			throw new EntityRetrievalException("Cannot update entity with id " + dto.getId() + ". Entity does not exist.");
+		}
+		
+		entity.setDataDate(new Date());
+		if(dto.getJsonDataObject() != null) {
+			entity.setJsonDataObject(dto.getJsonDataObject());
+		}
+		
+		if(dto.getTypeOfStatId() != null) {
+			entity.setTypeOfStatId(dto.getTypeOfStatId());
+		}
+		
+		if(dto.getLastModifiedUser() != null) {
+			entity.setLastModifiedUser(dto.getLastModifiedUser());
+		} else {
+			entity.setLastModifiedUser(Util.getCurrentUser().getId());
+		}		
+		
+		if(dto.getLastModifiedDate() != null) {
+			entity.setLastModifiedDate(dto.getLastModifiedDate());
+		} else {
+			entity.setLastModifiedDate(new Date());
+		}
+			
+		update(entity);
+		return new ChartDataDTO(entity);
+	}
+	
+	public ChartDataDTO getById(Long acbId) throws EntityRetrievalException{
+		ChartDataEntity entity = this.getEntityById(acbId);
+		
+		ChartDataDTO dto = null;
+		if(entity != null) {
+			dto = new ChartDataDTO(entity);
+		}
+		return dto;
+		
+	}
+
+	public List<ChartDataDTO> findAllData(){
+
+		List<ChartDataEntity> entities = getAllDataEntities();
+		List<ChartDataDTO> acbs = new ArrayList<>();
+
+		for (ChartDataEntity entity : entities) {
+			ChartDataDTO acb = new ChartDataDTO(entity);
+			acbs.add(acb);
+		}
+		return acbs;
+	}
+	
+	public List<ChartDataStatTypeDTO> findAllTypes(){
+
+		List<ChartDataStatTypeEntity> entities = getAllTypeEntities();
+		List<ChartDataStatTypeDTO> acbs = new ArrayList<>();
+
+		for (ChartDataStatTypeEntity entity : entities) {
+			ChartDataStatTypeDTO acb = new ChartDataStatTypeDTO(entity);
+			acbs.add(acb);
+		}
+		return acbs;
+	}
+	
+	public List<ChartDataEntity> getAllDataEntities() {
+		
+		List<ChartDataEntity> result;
+		
+		result = entityManager.createQuery( "from ChartDataEntity", ChartDataEntity.class).getResultList();
+
+		return result;
+	}
+	
+	public List<ChartDataStatTypeEntity> getAllTypeEntities() {
+		
+		List<ChartDataStatTypeEntity> result;
+		
+		result = entityManager.createQuery( "from ChartDataStatTypeEntity", ChartDataStatTypeEntity.class).getResultList();
+
+		return result;
+	}
+
+	public ChartDataEntity getEntityById(Long entityId) throws EntityRetrievalException {
+
+		ChartDataEntity entity = null;
+
+		String queryStr = "from ChartDataEntity "
+				+ "where (chart_data_id = :entityid)";
+		Query query = entityManager.createQuery(queryStr, ChartDataEntity.class );
+		query.setParameter("entityid", entityId);
+		List<ChartDataEntity> result = query.getResultList();
+
+		if (result.size() > 1){
+			throw new EntityRetrievalException("Data error. Duplicate certificaiton body id in database.");
```

Comment:
Typo on `certificaiton`; should be `certification`

Reply:
I have been pondering adding code to the update call because the tight coupling of putting code in there to fix all the statistics was not sitting well. Especially as we talked about adding other types of statistics that might be affected by various other things being updated. Can we implement this in a different way using a custom Spring event listener? Here's an example tutorial although we don't do this anywhere else so I don't know for sure that it will work with our versions of everything. https://www.tutorialspoint.com/spring/custom_events_in_spring.htm Basically, the CP Controller will become a class that can broadcast events and when they broadcast a custom ListingUpdatedEvent there will be a ChartDataUpdateHandler that listens for that event (and other events in the future) and handles it appropriately by going back in time and updating all that chart data. If that programming pattern works out we might be able to apply it to some other things, too.

Reply:
That sounds promising. Definitely don't like the high-coupling issue. @brianlindsy does that make sense to you? Would you like some of @kekey1 's help with this?

I know we're running way long on the task, and feel like we should double-down on this one and get it done